Date Rounding

spurll opened the pull request:
Adds methods to allow rounding `Date` and `DateTime` values to a specified resolution (1 month, 15 minutes, etc.) with `floor`, `ceil`, and `round`.

While we at Invenia believe that this functionality is useful and appropriate for base Julia, it could also be put in an external package in [JuliaTime](https://github.com/JuliaTime) if the community feels otherwise.

spurll committed:
Added floor, ceil, round for Date and DateTime.

spurll committed:
Refactoring and documentation for date rounding.

spurll committed:
Added PR reference in NEWS.md.

quinnj commented:
We already have `trunc`, so I think this makes sense.

StefanKarpinski commented:
Yes, huge 👍 to this. Great PR, @spurll

tkelman commented on doc/manual/dates.rst:
```diff
@@ -367,4 +367,95 @@ Periods are a human view of discrete, sometimes irregular durations of time. Con
+
+This means that rounding seconds, minutes, hours, or years (because the ISO 8601
+specification includes a year zero) to an even multiple will result in that field having
+an even value, while rounding to an even multiple of months will result in that field
```
what's the difference here? rounding to an even multiple (of what) results in even, when rounding to an even multiple of months results in odd?

spurll replied:
Sorry, I think I worded that poorly. It should read, "This means that rounding a `DateTime` to an even multiple of seconds, minutes, ... will result in that field having an even value in the resulting `DateTime`, while rounding a `DateTime` to an even multiple of months will result in the months field having an odd value."

I'm not sure if the use case would be sufficiently common for it to warrant comment, but I figured somebody might wonder why rounding a `Date`(`Time`) to `Month(2)` returns an odd-numbered month instead of an even-numbered month (as might be expected).

omus commented on base/dates/rounding.jl:
```diff
@@ -0,0 +1,173 @@
+epochms2datetime(i) = DateTime(UTM(DATETIMEEPOCH + Int64(i)))
+
+"""
+    date2epochdays(dt::DateTime) -> Int64
```
Should be `::Date`

spurll replied:
Good catch. I'll fix that in the morning.

spurll replied:
Also fixed a similar issue with the docstring for `epochdays2date`.

omus commented on test/dates/rounding.jl:
```diff
@@ -0,0 +1,150 @@
+@test Dates.date2epochdays(Dates.Date(2016, 1, 1)) == 736329
+@test Dates.datetime2epochms(Dates.DateTime(-1, 12, 31)) == -86400000
+@test Dates.datetime2epochms(Dates.DateTime(0, 1, 1)) == 0
+@test Dates.datetime2epochms(Dates.DateTime(2016, 1, 1)) == 736329 * 86400000
```
Integer overflow issue on 32-bit

spurll committed:
Bugfix for date rounding test in 32-bit. Doc clarifications.

spurll commented:
Hm. Of the three Travis jobs, one was successful (Linux 32), one ran out of storage space (Linux 64), and one was terminated because it took too long (OS X).

tlycken commented:
Nice!

Will this also be able to handle when the provided strings are overspecified? See http://stackoverflow.com/questions/37971649/parsing-zulu-dates-in-julia-throws-inexacterror for an example of what I mean.

spurll commented:
@tlycken: That behaviour will not be affected by this PR. The `round`/`floor`/`ceil` functions operate on existing `Date` and `DateTime` values; how strings are parsed into `DateTime`s is another matter.

spurll committed:
Round to nonpositive resolution throws DomainError

Throw DomainError on rounding to an invalid (non-positive) resolution
Clean up test cases for rounding dates that don't need rounding
Add test cases for rounding to invalid (non-positive) resolutions

kshyatt commented:
Hi @spurll, can you rebase to make this mergeable?

spurll commented:
Absolutely. Doing that now. Just had to add a quick fix to disallow rounding to non-positive `Period` values.

iamed2 commented:
This is good to go now; would be nice to get it in for 0.5.
